### PR TITLE
Add serial recipe

### PR DESCRIPTION
**Pull request by anton-danielsson**
Specify library name and version:  **serial/1.2.1**

Add recipe for https://github.com/wjwwood/serial

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/serial/all/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.5)
+project(cmake_wrapper)
+
+include(conanbuildinfo.cmake)
+conan_basic_setup()
+
+# Work-around for https://github.com/wjwwood/serial/issues/135
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR})
```
Comment (prince-chrismc) [inline recipes/serial/all/CMakeLists.txt:8]: Could you replace this with the content of the separate file?

File: recipes/serial/all/test_package/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.1.0)
+project(test_package CXX)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup(TARGETS)
+
+# Test components
+find_package(serial REQUIRED)
+add_executable(${PROJECT_NAME} test_package.cpp)
+target_link_libraries(${PROJECT_NAME} serial::serial)
```
Comment (prince-chrismc) [inline recipes/serial/all/test_package/CMakeLists.txt:10]: Some 💄 

```suggestion
cmake_minimum_required(VERSION 3.1)
project(test_package CXX)

include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
conan_basic_setup(TARGETS)

find_package(serial REQUIRED)

add_executable(${PROJECT_NAME} test_package.cpp)
target_link_libraries(${PROJECT_NAME} serial::serial)
```

File: recipes/serial/all/test_package/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        if not tools.cross_building(self):
+            bin_path = os.path.join("bin", "test_package")
+            self.run("%s -s" % bin_path, run_environment=True)
```
Comment (prince-chrismc) [inline recipes/serial/all/test_package/conanfile.py:16]: Command line is not read from

```suggestion
            self.run(bin_path, run_environment=True)
```

Comment (anton-danielsson) [reply recipes/serial/all/test_package/conanfile.py:16]: Thx, copy-paste stuff :)

File: recipes/serial/all/CMakeLists.txt (hunk 1 above)
Comment (anton-danielsson) [reply recipes/serial/all/CMakeLists.txt:8]: The idea was to make sure this line "works": https://github.com/wjwwood/serial/blob/master/CMakeLists.txt#L5
But you're right, it should be possible to inline it and set `catkin_FOUND` manually.
I will try that.

Comment (anton-danielsson) [reply recipes/serial/all/CMakeLists.txt:8]: No, I can't seem to trick CMake.
The `find_package(catkin REQUIRED)` fails even if I set `set(catkin_FOUND true)`...

**Review by prince-chrismc — APPROVED** [review]

**Comment by anton-danielsson** [comment]
Could we merge this?

**Comment by prince-chrismc** [comment]
If anyone gets a chance, this PR has been waiting a week for a second review, love to get more 👀 on it please 🙏 

@uilianries @SSE4 @madebr @SpaceIm @ericLemanissier @Croydon @intelligide @theirix @mathbunnyru

**Review by intelligide — APPROVED** [review]

**Review by madebr — APPROVED** [review]

File: recipes/serial/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    description = "Cross-platform library for interfacing with rs-232 serial like ports"
+    topics = ("serial", "rs-232", "com")
+    homepage = "http://wjwwood.io/serial/"
+    url = "https://github.com/conan-io/conan-center-index"
+    license = "MIT"
+    exports_sources = ["CMakeLists.txt", "Findcatkin.cmake"]
+    generators = "cmake"
+    settings = "os", "arch", "compiler", "build_type"
+    options = {"fPIC": [True, False]}
+    default_options = {"fPIC": True}
```
Comment (SpaceIm) [inline recipes/serial/all/conanfile.py:17]: Why no `shared` option? Upstream CMakeLists doesn't force STATIC.

File: recipes/serial/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def package(self):
+        self.copy("README.md", dst="licenses", src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
+
+    def package_info(self):
+        self.cpp_info.libs = tools.collect_libs(self)
```
Comment (SpaceIm) [inline recipes/serial/all/conanfile.py:51]:
```suggestion
        self.cpp_info.libs = ["serial"]
```

File: recipes/serial/all/CMakeLists.txt (hunk 1 above)
Comment (SpaceIm) [reply recipes/serial/all/CMakeLists.txt:8]: this line seems to be useless.

Comment (SpaceIm) [reply recipes/serial/all/CMakeLists.txt:8]: Honestly I would just patch upstream CMakeLists, the last release is 6 years old.

File: recipes/serial/all/conanfile.py (hunk 4 above)
Comment (anton-danielsson) [reply recipes/serial/all/conanfile.py:17]: Correct! I will fix!

Comment (anton-danielsson) [reply recipes/serial/all/conanfile.py]: Fixed in https://github.com/conan-io/conan-center-index/pull/4943/commits/f0869cffa133f40cded38b8bab4bf859e2c176d5

File: recipes/serial/all/CMakeLists.txt (hunk 1 above)
Comment (anton-danielsson) [reply recipes/serial/all/CMakeLists.txt:8]: Maybe... I think the Findcatkin.cmake is easier though.
And the installation can also be fixed without a patch. Like this: https://github.com/conan-io/conan-center-index/commit/f0869cffa133f40cded38b8bab4bf859e2c176d5#diff-196a5863890619f16fe188fdc1392c8fffe603fa16a60e515ba5211615ae54d9R12

I also want to add that I hope that the upstream repo will create a new release.
As you say the latest release is 6 years old, but the latest commit is just 1 years old :)
If I don't get a response we could maybe add a new version like `1.2.2-cbcca7c8`?
Download using: https://github.com/wjwwood/serial/archive/cbcca7c8.tar.gz

Comment (SpaceIm) [reply recipes/serial/all/CMakeLists.txt:8]: By fixing installation, I meant install destination of dll.

Comment (anton-danielsson) [reply recipes/serial/all/CMakeLists.txt:8]: Yes, me too. By adding a new install command it's fixed.
That was the plan at least.

Comment (anton-danielsson) [reply recipes/serial/all/CMakeLists.txt:8]:
```cmake
add_subdirectory(source_subfolder)

install(TARGETS serial)
```

File: recipes/serial/all/conanfile.py [hunk 6]
```diff
@@ -35,6 +45,8 @@ def _configure_cmake(self):
         if self._cmake:
             return self._cmake
         self._cmake = CMake(self)
+        if self.settings.os == "Windows" and self.options.shared:
+            self._cmake.definitions["CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS"] = True
```
Comment (prince-chrismc) [inline recipes/serial/all/conanfile.py:49]: We tend to keep this in the `cmake_wrapper` if you can please move it

Comment (anton-danielsson) [reply recipes/serial/all/conanfile.py:49]: Sure! I guess I just happened to pick the wrong way :)

**Review by SpaceIm — APPROVED** [review]
LGTM.
As I said, I would prefer a simple patch instead of the workaround in CMake wrapper + extra Findcatklin.cmake file, but the recipe is valid.

**Review by madebr — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Comment by anton-danielsson** [comment]
Not sure why license/cla is yellow...

**Review by intelligide — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**